File: packages/vault-item-types/src/catalogue.ts
/**
 * The closed field-type catalogue (ADR 0087 §2).
 *
 * A definition names a field type; it never describes one. That indirection —
 * Keeper's `$ref`, restated — is what keeps a definition inert: a manifest
 * that can only name behaviours cannot introduce one. Adding an entry here is
 * a platform change reviewed like any other; adding an *item type* is not.
 *
 * Every value in the vault is text, because the base native secret
 * (`sealed_store::Entry`) is text and every type projects onto it (§3). A
 * record-shaped field is a fixed set of named text parts, not an open object.
 */

export type FieldValueKind = "scalar" | "record";

export type FieldPart = {
  readonly id: string;
  readonly label: string;
  /** A concealed part makes the whole field concealed for the preview rule. */
  readonly concealed: boolean;
};

export type FieldTypeSpec = {
  readonly id: FieldTypeId;
  readonly valueKind: FieldValueKind;
  /**
   * True when the value must never render without a reveal gesture. Drives
   * §5's refusal of concealed fields in `subtitle`, `search`, and defaults.
   */
  readonly concealed: boolean;
  /** Renders as a text area rather than a single line. */
  readonly multiline: boolean;
  /** Present only on record shapes; the complete, ordered part list. */
  readonly parts: readonly FieldPart[];
};

export type FieldTypeId =
  | "string"
  | "multiline"
  | "email"
  | "url"
  | "number"
  | "boolean"
  | "date"
  | "month-year"
  | "country"
  | "select"
  | "phone"
  | "concealed"
  | "password"
  | "pin"
  | "key-material"
  | "totp"
  | "address"
  | "person-name"
  | "host-port"
  | "security-question"
  | "payment-card"
  | "key-pair";

function scalar(
  id: FieldTypeId,
  concealed = false,
  multiline = false,
): FieldTypeSpec {
  return { id, valueKind: "scalar", concealed, multiline, parts: [] };
}

function record(id: FieldTypeId, parts: readonly FieldPart[]): FieldTypeSpec {
  return {
    id,
    valueKind: "record",
    concealed: parts.some((part) => part.concealed),
    multiline: false,
    parts,
  };
}

function part(id: string, label: string, concealed = false): FieldPart {
  return { id, label, concealed };
}

export const FIELD_TYPES = {
  string: scalar("string"),
  multiline: scalar("multiline", false, true),
  email: scalar("email"),
  url: scalar("url"),
  number: scalar("number"),
  boolean: scalar("boolean"),
  date: scalar("date"),
  "month-year": scalar("month-year"),
  country: scalar("country"),
  select: scalar("select"),
  phone: scalar("phone"),
  concealed: scalar("concealed", true),
  password: scalar("password", true),
  pin: scalar("pin", true),
  "key-material": scalar("key-material", true, true),
  totp: scalar("totp", true),
  address: record("address", [
    part("street1", "Street"),
    part("street2", "Street 2"),
    part("city", "City"),
    part("state", "State or province"),
    part("postalCode", "Postal code"),
    part("country", "Country"),
  ]),
  "person-name": record("person-name", [
    part("first", "First"),
    part("middle", "Middle"),
    part("last", "Last"),
  ]),
  "host-port": record("host-port", [
    part("host", "Host"),
    part("port", "Port"),
  ]),
  "security-question": record("security-question", [
    part("question", "Question"),
    part("answer", "Answer", true),
  ]),
  "payment-card": record("payment-card", [
    part("number", "Number", true),
    part("expiry", "Expires"),
    part("code", "Security code", true),
  ]),
  "key-pair": record("key-pair", [
    part("publicKey", "Public key"),
    part("privateKey", "Private key", true),
  ]),
} satisfies Readonly<Record<FieldTypeId, FieldTypeSpec>>;

/** Written out rather than derived so the union and the table cannot drift
    apart silently; a test holds this list against `FIELD_TYPES`. */
export const FIELD_TYPE_IDS: readonly FieldTypeId[] = [
  "string",
  "multiline",
  "email",
  "url",
  "number",
  "boolean",
  "date",
  "month-year",
  "country",
  "select",
  "phone",
  "concealed",
  "password",
  "pin",
  "key-material",
  "totp",
  "address",
  "person-name",
  "host-port",
  "security-question",
  "payment-card",
  "key-pair",
];

export function isFieldTypeId(value: string): value is FieldTypeId {
  return FIELD_TYPE_IDS.some((candidate) => candidate === value);
}

export function fieldTypeSpec(id: FieldTypeId): FieldTypeSpec {
  return FIELD_TYPES[id];
}

/** True when a value of this field type must never render un-revealed. */
export function isConcealedFieldType(id: FieldTypeId): boolean {
  return FIELD_TYPES[id].concealed;
}

File: crates/vault-item-types/src/catalogue.rs
//! The closed field-type catalogue (ADR 0087 §2).
//!
//! Mirrors `packages/vault-item-types/src/catalogue.ts` entry for entry. A
//! definition names a field type; it never describes one, which is what keeps
//! a manifest inert. Adding an entry here is a platform change, and it must
//! land in both languages at once — `tests/conformance.rs` fails otherwise.

use serde::{Deserialize, Serialize};

/// A named text part of a record-shaped field.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct FieldPart {
    pub id: &'static str,
    pub label: &'static str,
    /// A concealed part makes the whole field concealed for the preview rule.
    pub concealed: bool,
}

const fn part(id: &'static str, label: &'static str) -> FieldPart {
    FieldPart {
        id,
        label,
        concealed: false,
    }
}

const fn secret_part(id: &'static str, label: &'static str) -> FieldPart {
    FieldPart {
        id,
        label,
        concealed: true,
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum FieldShape {
    Scalar,
    Record,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "kebab-case")]
pub enum FieldTypeId {
    String,
    Multiline,
    Email,
    Url,
    Number,
    Boolean,
    Date,
    MonthYear,
    Country,
    Select,
    Phone,
    Concealed,
    Password,
    Pin,
    KeyMaterial,
    Totp,
    Address,
    PersonName,
    HostPort,
    SecurityQuestion,
    PaymentCard,
    KeyPair,
}

/// Every catalogue entry, in the same order as the TypeScript table.
pub const FIELD_TYPE_IDS: [FieldTypeId; 22] = [
    FieldTypeId::String,
    FieldTypeId::Multiline,
    FieldTypeId::Email,
    FieldTypeId::Url,
    FieldTypeId::Number,
    FieldTypeId::Boolean,
    FieldTypeId::Date,
    FieldTypeId::MonthYear,
    FieldTypeId::Country,
    FieldTypeId::Select,
    FieldTypeId::Phone,
    FieldTypeId::Concealed,
    FieldTypeId::Password,
    FieldTypeId::Pin,
    FieldTypeId::KeyMaterial,
    FieldTypeId::Totp,
    FieldTypeId::Address,
    FieldTypeId::PersonName,
    FieldTypeId::HostPort,
    FieldTypeId::SecurityQuestion,
    FieldTypeId::PaymentCard,
    FieldTypeId::KeyPair,
];

const ADDRESS_PARTS: [FieldPart; 6] = [
    part("street1", "Street"),
    part("street2", "Street 2"),
    part("city", "City"),
    part("state", "State or province"),
    part("postalCode", "Postal code"),
    part("country", "Country"),
];

const PERSON_NAME_PARTS: [FieldPart; 3] = [
    part("first", "First"),
    part("middle", "Middle"),
    part("last", "Last"),
];

const HOST_PORT_PARTS: [FieldPart; 2] = [part("host", "Host"), part("port", "Port")];

const SECURITY_QUESTION_PARTS: [FieldPart; 2] = [
    part("question", "Question"),
    secret_part("answer", "Answer"),
];

const PAYMENT_CARD_PARTS: [FieldPart; 3] = [
    secret_part("number", "Number"),
    part("expiry", "Expires"),
    secret_part("code", "Security code"),
];

const KEY_PAIR_PARTS: [FieldPart; 2] = [
    part("publicKey", "Public key"),
    secret_part("privateKey", "Private key"),
];

impl FieldTypeId {
    /// The wire spelling, as it appears in a definition's `type`.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::String => "string",
            Self::Multiline => "multiline",
            Self::Email => "email",
            Self::Url => "url",
            Self::Number => "number",
            Self::Boolean => "boolean",
            Self::Date => "date",
            Self::MonthYear => "month-year",
            Self::Country => "country",
            Self::Select => "select",
            Self::Phone => "phone",
            Self::Concealed => "concealed",
            Self::Password => "password",
            Self::Pin => "pin",
            Self::KeyMaterial => "key-material",
            Self::Totp => "totp",
            Self::Address => "address",
            Self::PersonName => "person-name",
            Self::HostPort => "host-port",
            Self::SecurityQuestion => "security-question",
            Self::PaymentCard => "payment-card",
            Self::KeyPair => "key-pair",
        }
    }

    #[must_use]
    pub const fn shape(self) -> FieldShape {
        match self {
            Self::Address
            | Self::PersonName
            | Self::HostPort
            | Self::SecurityQuestion
            | Self::PaymentCard
            | Self::KeyPair => FieldShape::Record,
            _ => FieldShape::Scalar,
        }
    }

    /// The complete, ordered part list; empty for every scalar shape.
    #[must_use]
    pub const fn parts(self) -> &'static [FieldPart] {
        match self {
            Self::Address => &ADDRESS_PARTS,
            Self::PersonName => &PERSON_NAME_PARTS,
            Self::HostPort => &HOST_PORT_PARTS,
            Self::SecurityQuestion => &SECURITY_QUESTION_PARTS,
            Self::PaymentCard => &PAYMENT_CARD_PARTS,
            Self::KeyPair => &KEY_PAIR_PARTS,
            _ => &[],
        }
    }

    /// True when the value must never render without a reveal gesture. A
    /// record shape is concealed when any one of its parts is.
    #[must_use]
    pub const fn concealed(self) -> bool {
        matches!(
            self,
            Self::Concealed
                | Self::Password
                | Self::Pin
                | Self::KeyMaterial
                | Self::Totp
                | Self::SecurityQuestion
                | Self::PaymentCard
                | Self::KeyPair
        )
    }

    #[must_use]
    pub const fn multiline(self) -> bool {
        matches!(self, Self::Multiline | Self::KeyMaterial)
    }
}
